## Programming part of Homework 2 (Data Structures, Fall 2025)

## Name: 曾靖諺
## Student ID Number: 113820032

### Programming problem 1
Consider the problem to compute the product of two positive integers, $n$ and $m$, using only addition and subtraction.
This programming problem is to consider the recursive approach in above problem and asks you to complete the following tasks.
1. Implement the approach as a function named as `product_rec` using Python.
2. Please have an iterative version for the approach and write a function for this version with function name `product_ite`.
3. Compare these two function with the same input in terms of running time and write what you have observed.

**Note** that we will use an in-built python library **`timeit`** and the module function **`timeit.timeit()`** for measuring the running time.
For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.

In [3]:
def product_rec(n, m):
    if (m == 1):
        return n
    else:
        return n + product_rec(n, m-1)

print("The product of 4 and 7 is", product_rec(4, 7))

The product of 4 and 7 is 28


In [4]:
def product_ite(n, m):
    result = 0

    for i in range(m):
        result += n

    return result

print("The product of 4 and 7 is", product_ite(4, 7))

The product of 4 and 7 is 28


In [10]:
#---------------------------------------------------------timeit
import timeit as ti

#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")
tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_rec(80, 700)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(10000):
    product_ite(80, 700)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值
print("Recursion 80*700=", product_rec(80, 700),"Iteration 80*700=", product_ite(80, 700))
print("Recursive approach:", t_rec)
print("iterative approach:", t_ite)

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")
t_rec = ti.timeit("product_rec(80, 700)", setup="from __main__ import product_rec", number = 10000)#timeit(函數名稱,執行次數)
t_ite = ti.timeit("product_ite(80, 700)", setup="from __main__ import product_ite", number = 10000)

print("Recursion 80*700=", product_rec(80, 700),"Iteration 80*700=", product_ite(80, 700))
print("Recursive approach:", t_rec)
print("Iterative approach:", t_ite)

Running time comparison (Timer Method 1):
Recursion 80*700= 56000 Iteration 80*700= 56000
Recursive approach: 1.0916520999744534
iterative approach: 0.2828603000380099
Running time comparison (Timer Method 2):
Recursion 80*700= 56000 Iteration 80*700= 56000
Recursive approach: 1.080564699950628
Iterative approach: 0.29782050009816885


#### Please write down the observations you have after the comparisons:
##### Inputs (values of n) for five experiments
1.<br>40</br>
2.<br>50</br>
3.<br>60</br>
4.<br>70</br>
5.<br>80</br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.30737100006081164 | 0.2797376000089571 | 0.3030913999537006 | 0.2775906000752002 | 0.2828603000380099 |
|Recursion      | 1.0566684999503195 | 1.1072623000945896 | 1.1302462999010459 | 1.106996100046672 | 1.0916520999744534 |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.3293694999301806 | 0.3291188000002876 | 0.2869778999593109 | 0.2766126999631524 | 0.29782050009816885 |
|Recursion      | 1.1887150999391451 | 1.1373448999365792 | 1.1021798000438139 | 1.0854448999743909 | 1.080564699950628 |

*******************************************************************************************************

##### Observations
理論上遞迴(Recursion)與迭代(Iteration)方法的時間複雜度相同，皆為 O(n)。然而，實驗使用兩種方法測試，結果均顯示遞迴用時在1.05 ~ 1.19之間、迭代用時在0.27 ~ 0.33之間，差異約在3.2 ~ 4.3倍，這可能是因為遞迴法在每次呼叫時需要額外的函式呼叫，因此比迭代慢。

### Programming problem 2
Consider the problem of ***Chebyshev polynomials of the first kind***. Please use Python to implement the pseudo-code you provide for solving the problem. You should provide all the functions with the mentioned names, `chebyshev_recursive`, `chebyshev_memoreccu`, `chebyshev_addoutrec`, and `chebyshev_iterative` respectively. Then, perform a comparison on these functions by measuring the execution time.
1. Please have the recursive function, `chebyshev_recursive` to list all the polynomials with the recursive definition.
2. Please provide the memoized recursive function, `chebyshev_memoreccu` to list all the polynomials.
3. Please use the way`LinearFibonacci}` we introduced in class to have a linear recursive function `chebyshev_addoutrec`.
4. Please provide an iterative (no recursive call) function, `chebyshev_iterative`, which lists all the polynomials iteratively.
5. In order to have the experiments for comparisons, please execute all the functions with the same input many times and then measure the practical running timme in average. Please write down your observations on these four versions.

**Note** that we will use an in-built python library **`timeit`** and the module function **`timeit.timeit()`** for measuring the running time.
For the programming problems, please do some experiments by yourself to observe the running time and the number of recursive calls. You need to report what you have observed from the experiments when you submit your homework.


In [5]:
def chebyshev_recursive(x):
    if (x == 0):
        return [1]
    elif (x == 1):
        return [1, 0]
    else:
        temp1 = chebyshev_recursive(x - 1)
        temp2 = chebyshev_recursive(x - 2)
        result = list()

        for i in range(len(temp1)):
            result.append(temp1[i] * 2)

        result.append(0)

        offset = len(result) - len(temp2)
        
        for i in range(0, len(temp2)):
            result[i+offset] -= temp2[i]

        return result

if __name__ == "__main__":
    # demo: print first 10 polynomials
    for n in range(11):
         print(f"The coefficients of T_{n}(x) in list = {chebyshev_recursive(n)}")

The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [10]:
def chebyshev_memoreccu(x, m_table):
    if (m_table[x] is not None):
        return m_table[x]
    if (x == 0):
        result = [1]
    elif (x == 1):
        result = [1, 0]
    else:
        temp1 = chebyshev_memoreccu(x - 1, m_table)
        temp2 = chebyshev_memoreccu(x - 2, m_table)
        result = list()

        for i in range(len(temp1)):
            result.append(temp1[i] * 2)

        result.append(0)

        offset = len(result) - len(temp2)
        
        for i in range(0, len(temp2)):
            result[i+offset] -= temp2[i]

    m_table[x] = result
    return result

if __name__ == "__main__":
    # demo: print first 10 polynomials
    for n in range(11):
      m_table=[None] * (n+1)
      print(f"The coefficients of T_{n}(x) in list = {chebyshev_memoreccu(n, m_table)}")

The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [16]:
def chebyshev_addoutrec(x):
    def linear_recursive(x, prev_x1, prev_x2):
        if (x == 0):
            return prev_x2
        elif (x == 1):
            return prev_x1
        else:
            result = list()

            for i in range(len(prev_x1)):
                result.append(prev_x1[i] * 2) # T[x] = 2 * T[x-1]

            result.append(0) # T[x] = x * T[x]

            offset = len(result) - len(prev_x2)
            
            for i in range(len(prev_x2)):
                result[i+offset] -= prev_x2[i]  # T[x] = T[x] - T[x-2]

            return linear_recursive(x - 1, result, prev_x1)

    return linear_recursive(x, [1, 0], [1])
    
if __name__ == "__main__":
    # demo: print first 10 polynomials
    for n in range(11):
      print(f"The coefficients of T_{n}(x) in list = {chebyshev_addoutrec(n)}")

The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [17]:
def chebyshev_iterative(x):
    prev_n2 = [1]
    prev_n1 = [1, 0]

    memo = [prev_n2, prev_n1]

    if (x <= 1):
        return memo[x]
    else:
        for i in range(2, x + 1):
            temp1 = memo[i-1]
            temp2 = memo[i-2]
            result = list()
            for j in range(len(temp1)):
                result.append(temp1[j] * 2)
            result.append(0)
            offset = len(result) - len(temp2)
            for j in range(0, len(temp2)):
                result[j+offset] -= temp2[j]
            memo.append(result)
        return memo[x]

if __name__ == "__main__":
    # demo: print first 10 polynomials
    for n in range(11):
         print(f"The coefficients of T_{n}(x) in list = {chebyshev_iterative(n)}")

The coefficients of T_0(x) in list = [1]
The coefficients of T_1(x) in list = [1, 0]
The coefficients of T_2(x) in list = [2, 0, -1]
The coefficients of T_3(x) in list = [4, 0, -3, 0]
The coefficients of T_4(x) in list = [8, 0, -8, 0, 1]
The coefficients of T_5(x) in list = [16, 0, -20, 0, 5, 0]
The coefficients of T_6(x) in list = [32, 0, -48, 0, 18, 0, -1]
The coefficients of T_7(x) in list = [64, 0, -112, 0, 56, 0, -7, 0]
The coefficients of T_8(x) in list = [128, 0, -256, 0, 160, 0, -32, 0, 1]
The coefficients of T_9(x) in list = [256, 0, -576, 0, 432, 0, -120, 0, 9, 0]
The coefficients of T_10(x) in list = [512, 0, -1280, 0, 1120, 0, -400, 0, 50, 0, -1]


In [32]:
#---------------------------------------------------------timeit
import timeit as ti
n=5
#---------------------------------------------------------Method 1
print("Running time comparison (Timer Method 1):")

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(1000):
    chebyshev_iterative(n)
tend = ti.default_timer() #timer end計時器關閉
t_ite=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(1000):
    chebyshev_recursive(n)
tend = ti.default_timer() #timer end計時器關閉
t_rec=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(1000):
    m_table=[None] * (n+1)
    chebyshev_memoreccu(n, m_table)
tend = ti.default_timer() #timer end計時器關閉
t_mem=tend - tstart#開始與結束的差值

tstart = ti.default_timer() #timer start 計時器開啟
for i in range(1000):
    chebyshev_addoutrec(n)
tend = ti.default_timer() #timer end計時器關閉
t_add=tend - tstart#開始與結束的差值

print(f"Deriving T_{{{n}}}(x)")
print(f"{"Iterative approach":>21} :{t_ite:<20}")
print(f"{"Recursive approach":>21} :{t_rec:<20}")
print(f"{"Memoizerec approach":>21} :{t_mem:<20}")
print(f"{"Addoutrec approach":>21} :{t_add:<20}")

#---------------------------------------------------------Method 2
print("Running time comparison (Timer Method 2):")

def iterative_test():
    chebyshev_iterative(n)

def recursive_test():
    chebyshev_recursive(n)

def memoreccu_test():
    m_table=[None] * (n+1)
    chebyshev_memoreccu(n, m_table)

def addoutrec_test():
    chebyshev_addoutrec(n)

t_ite = ti.timeit("iterative_test()", setup="from __main__ import iterative_test", number = 1000)#timeit(函數名稱,執行次數)
t_rec = ti.timeit("recursive_test()", setup="from __main__ import recursive_test", number = 1000)
t_mem = ti.timeit("memoreccu_test()", setup="from __main__ import memoreccu_test", number = 1000)
t_add = ti.timeit("addoutrec_test()", setup="from __main__ import addoutrec_test", number = 1000)

print(f"Deriving T_{{{n}}}(x)")
print(f"{"Iterative approach":>21} :{t_ite:<20}")
print(f"{"Recursive approach":>21} :{t_rec:<20}")
print(f"{"Memoizerec approach":>21} :{t_mem:<20}")
print(f"{"Addoutrec approach":>21} :{t_add:<20}")

Running time comparison (Timer Method 1):
Deriving T_{5}(x)
   Iterative approach :0.0018090000376105309
   Recursive approach :0.002922200015746057
  Memoizerec approach :0.002110400004312396
   Addoutrec approach :0.002325400011613965
Running time comparison (Timer Method 2):
Deriving T_{5}(x)
   Iterative approach :0.0016920999623835087
   Recursive approach :0.0028789000352844596
  Memoizerec approach :0.0020825000246986747
   Addoutrec approach :0.0019887000089511275


#### Please write down the observations you have after the comparisons:
##### Inputs for five experiments
1.<br>5</br>
2.<br>10</br>
3.<br>15</br>
4.<br>20</br>
5.<br>25</br>
##### Results of comparisoms
*******************************************************************************************************
**Method 1**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.0018090000376105309 | 0.005104199983179569 | 0.010873799910768867 | 0.015889300033450127 | 0.02785720000974834 |
|Recursion      | 0.002325400011613965 | 0.009263799991458654 | 0.018902899930253625 | 0.030222499975934625 | 0.0501802999060601 |
|Binrecurs      | 0.002922200015746057 | 0.06223289994522929 | 0.8127438999945298 | 9.38927759998478 | 106.03389520000201 |

**Method 2**

| time          | time for exp1  | time for exp2  | time for exp3  | time for exp4  | time for exp5  | 
| -----------   | -------------- | -------------- | -------------- | -------------- | -------------- |
|Iteration      | 0.0016920999623835087 | 0.009294499992392957 | 0.011335500050336123 | 0.02950030006468296 | 0.04407880001235753 |
|Recursion      | 0.0019887000089511275 | 0.008930400013923645 | 0.017564699985086918 | 0.03947910002898425 | 0.04494440008420497 |
|Binrecurs      | 0.0028789000352844596 | 0.07045300002209842 | 0.8930961999576539 | 9.468407800071873 | 105.22646580007859 |

*******************************************************************************************************

##### Observations
由結果可以看出Binrecurs(chebyshev_recursive)最慢，因為他每一次計算T_n都要呼叫T_(n-1)、T_(n-2)而導致計算量呈指數性增長。而不出乎意料的，迭代 Iteration(chebyshev_iterative)在所有測試結果中幾乎都是最快的，即便他與遞迴 Recursion(chebyshev_addoutrec)都是O(n^2)，然而也跟上一題一樣因為遞迴法在每次呼叫時需要額外的函式呼叫，因此比迭代慢一些。線性遞迴(Addoutrec)相比迭代(Iteration)慢一些，與記憶遞迴(Memoreccu)相近，但仍比Binrecurs快得多，因為他相同的T_n只計算一遍，避免了重複計算。